Zend/zend_alloc.c

```diff
@@ -3009,6 +3091,206 @@ ZEND_API void zend_mm_get_custom_handlers(zend_mm_heap *heap,
 #endif
 }
 
+/*
+ * Use this function to register observers to the ZendMM. Make sure to call this
+ * function only during PHP module startup phase and not later. Observers will
+ * not be called directly after registering them, but only after
+ * `zend_mm_observers_startup()` was run, which will take place after the PHP
+ * module startup phase.
+ */
+ZEND_API zend_mm_observer* zend_mm_observer_register(
+	void (*malloc)(size_t, void * ZEND_FILE_LINE_DC ZEND_FILE_LINE_ORIG_DC),
+	void (*free)(void * ZEND_FILE_LINE_DC ZEND_FILE_LINE_ORIG_DC),
+	void (*realloc)(void *, size_t, void * ZEND_FILE_LINE_DC ZEND_FILE_LINE_ORIG_DC)
+) {
+#if ZEND_MM_CUSTOM
+	zend_mm_observer *node = pemalloc(sizeof(zend_mm_observer), 1);
+	node->malloc = malloc;
+	node->free = free;
+	node->realloc = realloc;
+	node->next = NULL;
+	if (zend_mm_observers == NULL) {
+		zend_mm_observers = node;
+		return zend_mm_observers;
+	}
+	zend_mm_observer *current = zend_mm_observers;
+	while (current->next != NULL) {
+		current = current->next;
+	}
+	current->next = node;
+	return node;
+#else
+	return NULL;
+#endif
+}
+
+/*
+ * This function will activate the global registered observers for the current
+ * heap. After this has been run, observering functions will be called.
+ */
+void zend_mm_observers_startup(zend_mm_heap *heap)
+{
+#if ZEND_MM_CUSTOM
+	if (heap == NULL) {
+		heap = AG(mm_heap);
+	}
+	zend_mm_observer *current_observer = zend_mm_observers;
+	if (current_observer == NULL) {
+		// no observers installed
+		return;
+	}
+	// copy observers into heap
+	zend_mm_heap_observer *node = NULL;
+	zend_mm_heap_observer *current_heap_observer = NULL;
+	while (current_observer != NULL) {
+		node = pemalloc(sizeof(zend_mm_heap_observer), 1);
```

Comment:
Do we need two separate structs here? May we not merge `zend_mm_observer` and `zend_mm_heap_observer`, register them to `zend_mm_observers` and them copy the whole pointer to `zend_mm_heap.observers`? This would avoid an additional indirection and seems more straightforward. Maybe I'm missing something though.

This only seems necessary if we register multiple heaps that need their own observers. But I don't think that's the case?